packages/react/src/components/UIShell/HeaderContainer.tsx

```diff
@@ -0,0 +1,82 @@
+/**
+ * Copyright IBM Corp. 2016, 2023
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import PropTypes from 'prop-types';
+import React, { useState, useCallback } from 'react';
+
+interface HeaderContainerRenderProps {
+  isSideNavExpanded: boolean;
+  onClickSideNavExpand: () => void;
+}
+
+interface HeaderContainerProps {
+  isSideNavExpanded?: boolean;
+  render: React.ComponentType<HeaderContainerRenderProps>;
+}
+
+export default function HeaderContainer({
+  render: Children,
+  isSideNavExpanded = false,
+}: HeaderContainerProps) {
+  //state for expandable sidenav
+  const [isSideNavExpandedState, setIsSideNavExpandedState] =
+    useState(isSideNavExpanded);
+
+  const handleHeaderMenuButtonClick = useCallback(() => {
+    setIsSideNavExpandedState(
+      (prevIsSideNavExpanded) => !prevIsSideNavExpanded
+    );
+  }, [setIsSideNavExpandedState]);
+
+  return (
+    <Children
+      isSideNavExpanded={isSideNavExpandedState}
+      onClickSideNavExpand={handleHeaderMenuButtonClick}
+    />
+  );
+}
+
+HeaderContainer.propTypes = {
+  /**
+   * Optionally provide a custom class name that is applied to the underlying <header>
+   */
+  isSideNavExpanded: PropTypes.bool,
+
+  /**
+   * A function that is passed two parameters, `isSideNavExpanded`, and
+   * `onClickSideNavExpand`, which can then be used the returned components. For
+   * example:
```

Comment:
The comment looks good, I don't think we need the full example though as it is displayed in the stories.